jms/src/main/scala/akka/stream/alpakka/jms/Jms.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.jms
+
+import javax.jms
+import javax.jms._
+
+import akka.stream.stage.{ GraphStageLogic, StageLogging }
+
+trait JmsSettings {
+  def connectionFactory: ConnectionFactory
+  def destination: Option[Destination]
+  def credentials: Option[Credentials]
+}
+
+sealed trait Destination
+case class Topic(name: String) extends Destination
+case class Queue(name: String) extends Destination
+
+object JmsSourceSettings {
+
+  def create(connectionFactory: ConnectionFactory) = JmsSourceSettings(connectionFactory)
+
+}
+
+case class JmsSourceSettings(connectionFactory: ConnectionFactory,
+                             destination: Option[Destination] = None,
+                             credentials: Option[Credentials] = None,
+                             bufferSize: Int = 100)
+    extends JmsSettings {
+  def withCredential(credentials: Credentials) = copy(credentials = Some(credentials))
+  def withBufferSize(size: Int) = copy(bufferSize = size)
+  def withQueue(name: String) = copy(destination = Some(Queue(name)))
+  def withTopic(name: String) = copy(destination = Some(Topic(name)))
+}
+
+object JmsSinkSettings {
+
+  def create(connectionFactory: ConnectionFactory) = JmsSinkSettings(connectionFactory)
+
+}
+
+case class JmsSinkSettings(connectionFactory: ConnectionFactory,
+                           destination: Option[Destination] = None,
+                           credentials: Option[Credentials] = None)
+    extends JmsSettings {
+  def withCredential(credentials: Credentials) = copy(credentials = Some(credentials))
+  def withQueue(name: String) = copy(destination = Some(Queue(name)))
+  def withTopic(name: String) = copy(destination = Some(Topic(name)))
+}
+
+case class Credentials(username: String, password: String)
+
+trait JmsConnector { this: GraphStageLogic with StageLogging =>
+
+  private[jms] var jmsConnection: Option[Connection] = None
+  private[jms] var jmsSession: Option[Session] = None
+  private[jms] var jmsDestination: Option[jms.Destination] = None
+
+  def jmsSettings: JmsSettings
+
+  def fail = getAsyncCallback[JMSException](e => failStage(e))
+
+  def openQueue(): Unit = {
+    val factory: ConnectionFactory = jmsSettings.connectionFactory
+    try {
+      jmsConnection = Option(jmsSettings.credentials match {
+        case Some(Credentials(username, password)) => factory.createConnection(username, password)
+        case _ => factory.createConnection()
+      })
```

Comment:
If the entire stage runs in the blocking dispatcher it is easier to write as it can block whenever it wants without problems, it will however always introduce an asynchronous boundary, as the other stages in streams it is used in will not run on that same dispatcher (so same as `.async`), if just some parts of the internal uses future/async-callback you will have the boundary sort of inside the stage.

What is best is different from case to case, but as you already deal with a separate thread doing interacting with JMS and then handing over the elements to the stage and vice versa I think just running the small parts on a suitable dispatcher is a good solution.